security/pkg/nodeagent/sds/server.go

```diff
@@ -256,14 +263,21 @@ func (s *Server) initGatewaySdsService(options *Options) error {
 	go func() {
 		sdsServiceLog.Info("Start SDS grpc server for ingress gateway proxy")
 		waitTime := time.Second
+
 		for i := 0; i < maxRetryTimes; i++ {
-			// Retry if Serve() fails
+			success := true
 			if err = s.grpcGatewayServer.Serve(s.grpcGatewayListener); err != nil {
 				sdsServiceLog.Errorf("SDS grpc server for ingress gateway proxy failed to start: %v", err)
+				success = false
+			}
+			if success {
+				if s.grpcGatewayListener, err = setUpUds(options.IngressGatewayUDSPath); err != nil {
+					sdsServiceLog.Errorf("SDS grpc server for ingress gateway proxy failed to set up UDS: %v", err)
+					success = false
+				}
 			}
-			s.grpcGatewayListener, err = setUpUds(options.IngressGatewayUDSPath)
-			if err != nil {
-				sdsServiceLog.Errorf("SDS grpc server for ingress gateway proxy failed to set up UDS: %v", err)
+			if success {
+				break
```

Comment:
I donot think the previous logic is right when setUpUds failed, the s.grpcWorkloadListener is nil, no need to call  s.grpcWorkloadServer.Serve.

Reply:
Yes, you are correct, thanks for your comment